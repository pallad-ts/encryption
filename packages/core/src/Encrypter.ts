import { Ciphertext } from "./Ciphertext";
import { KeyRing } from "@pallad/keyring";
import { TextEncoding } from "./TextEncoding";

export abstract class Encrypter {
    constructor(readonly keyRing: KeyRing) {}

    abstract encrypt(data: Buffer): Promise<Ciphertext>;

    async encryptFromString(data: string, encoding: TextEncoding): Promise<Ciphertext> {
        const buffer = Buffer.from(data, encoding);
        return this.encrypt(buffer);
    }

    abstract decrypt(data: Ciphertext): Promise<Buffer>;

    async decryptToString(data: Ciphertext, encoding: TextEncoding): Promise<string> {
        const decrypted = await this.decrypt(data);
        return Buffer.from(decrypted).toString(encoding);
    }
}
