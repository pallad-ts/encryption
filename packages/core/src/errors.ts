import { ErrorDescriptor, formatCodeFactory, Domain } from "@pallad/errors";
import { EncryptionError } from "./EncryptionError";

const code = formatCodeFactory("E_PALLAD_ENCRYPTION_%c");

export const ERRORS = new Domain().addErrorsDescriptorsMap({
    INVALID_KEYRING: ErrorDescriptor.useDefaultMessage(
        code(1),
        "Provided keyring is not the one configured by encrypter",
        EncryptionError
    ),
});
