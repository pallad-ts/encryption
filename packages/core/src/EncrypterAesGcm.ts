import { Ciphertext } from "./Ciphertext";
import { Encrypter } from "./Encrypter";
import { z } from "zod";
import { webcrypto } from "node:crypto";
import { KeyRing, KeyId } from "@pallad/keyring";
import { Range } from "@pallad/range";
import { left, right } from "@sweet-monads/either";
import { ERRORS } from "./errors";

const ALGORITHM = "AES-GCM";

const CRYPTO_KEY_MAP = new WeakMap<Buffer, Promise<CryptoKey>>();

const KEY_RING = Symbol("KeyRing");

function getCryptoKey(buffer: Buffer): Promise<CryptoKey> {
    let key = CRYPTO_KEY_MAP.get(buffer);
    if (!key) {
        key = webcrypto.subtle.importKey("raw", buffer, ALGORITHM, false, ["encrypt", "decrypt"]);
        CRYPTO_KEY_MAP.set(buffer, key);
    }
    return key;
}

export class EncrypterAesGcm extends Encrypter {
    readonly options: EncrypterAesGcm.Options;

    constructor(
        readonly keyRing: KeyRing,
        options?: EncrypterAesGcm.Options.FromUser
    ) {
        validateKeyRing(keyRing);
        super(keyRing);

        this.options = EncrypterAesGcm.OptionsSchema.parse({
            ivLength: 12,
            ...options,
        });
    }

    async encrypt(data: ArrayBuffer, keyId?: KeyId): Promise<Ciphertext> {
        const key = keyId ? this.keyRing.assertEntryById(keyId) : this.keyRing.getRandomKey();

        const [iv, cryptoKey] = await Promise.all([
            webcrypto.getRandomValues(Buffer.alloc(this.options.ivLength)),
            getCryptoKey(key.key.getValue()),
        ]);

        const encrypted = Buffer.from(
            await webcrypto.subtle.encrypt({ name: ALGORITHM, iv, tagLength: 128 }, cryptoKey, data)
        );
        return Ciphertext.schema.parse({ keyId: key.id, iv, encrypted });
    }

    async decrypt(data: Ciphertext): Promise<Buffer> {
        const { iv, encrypted } = data;
        const key = this.keyRing.assertKeyById(data.keyId);
        const cryptoKey = await getCryptoKey(key.getValue());
        return Buffer.from(await webcrypto.subtle.decrypt({ name: ALGORITHM, iv }, cryptoKey, encrypted));
    }

    static createKeyRing() {
        const keyRing = new KeyRing({
            keySize: Range.create(12),
            validation: validateKeyEntry,
        });

        (keyRing as any)[KEY_RING] = true;
        return keyRing;
    }
}

const ALLOWED_SIZES = new Set([128, 192, 256]);

function validateKeyEntry(entry: KeyRing.Entry) {
    const keySize = entry.key.getValue().length * 8;

    if (!ALLOWED_SIZES.has(keySize)) {
        return left(`Key size ${keySize} is not allowed`);
    }

    return right(entry);
}

function validateKeyRing(keyRing: KeyRing) {
    if ((keyRing as any)[KEY_RING] !== true) {
        throw ERRORS.INVALID_KEYRING.create();
    }
}

export namespace EncrypterAesGcm {
    export const OptionsSchema = z.object({
        ivLength: z.number().int().min(12).max(64),
    });
    export type Options = z.infer<typeof OptionsSchema>;

    export namespace Options {
        export type FromUser = Partial<Options>;
    }
}
