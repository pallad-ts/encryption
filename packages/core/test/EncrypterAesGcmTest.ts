import { EncrypterAesGcm } from "@src/EncrypterAesGcm";
import { randomBytes } from "crypto";
import { Encrypter } from "@src/Encrypter";
import { KeyRing } from "@pallad/keyring";
import "@pallad/errors-dev";
import { ERRORS } from "@src/errors";

describe("EncrypterAesGcm", () => {
    let encrypter: EncrypterAesGcm;
    beforeEach(() => {
        const keyRing = EncrypterAesGcm.createKeyRing();
        keyRing.addKey("key1", randomBytes(32));
        keyRing.addKey("key2", randomBytes(32));
        keyRing.addKey("key3", randomBytes(32));
        encrypter = new EncrypterAesGcm(keyRing);
    });

    async function assertEncodesAndDecodes(encrypter: Encrypter, input: string = "foobar") {
        const encrypted = await encrypter.encryptFromString(input, "utf8");
        const decrypted = await encrypter.decryptToString(encrypted, "utf8");
        expect(decrypted).toEqual(input);
    }

    it.each([["somerandomstring"], [JSON.stringify({ some: { random: "data" } })]])(
        "encodes and decodes data",
        input => {
            return assertEncodesAndDecodes(encrypter, input);
        }
    );

    describe("creating", () => {
        it("fails when using keyring that is not created from EncrypterAesGcm.createKeyRing", () => {
            const keyRing = new KeyRing();
            keyRing.addKey("key1", randomBytes(32));
            expect(() => new EncrypterAesGcm(keyRing)).toThrowErrorWithCode(ERRORS.INVALID_KEYRING);
        });

        it("keyring will not accept keys with invalid size", () => {
            const keyRing = EncrypterAesGcm.createKeyRing();

            expect(() => {
                keyRing.addKey("key_invalid", Buffer.alloc(50));
            }).toThrowErrorMatchingSnapshot();
        });
    });
});
