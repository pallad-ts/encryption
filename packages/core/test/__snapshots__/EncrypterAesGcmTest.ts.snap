// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EncrypterAesGcm creating keyring will not accept keys with invalid size 1`] = `"Key custom validation violated: Key size 400 is not allowed"`;
